pkg/plugins/queue/azqueue/azqueue.go

```diff
@@ -0,0 +1,287 @@
+// Azure requires a combination of their unique reference for a queue item (id) and a pop receipt (lease id)
+// to perform operations on the item, such as delete it from the queue.
+type AzureQueueItemLease struct {
+	// The ID of the queue item
+	// note: this is an id generated by Azure, it's not the user provided unique id.
+	ID string
+	// lease id, a new popReceipt is generated each time an item is dequeued.
+	PopReceipt string
+}
+
+// String - convert the item lease struct to a string, to be returned as a NitricTask LeaseID
+func (l *AzureQueueItemLease) String() (string, error) {
+	leaseID, err := json.Marshal(l)
+	return string(leaseID), err
+}
+
+// leaseFromString - Unmarshal a NitricTask Lease ID (JSON) to an AzureQueueItemLease
+func leaseFromString(leaseID string) (*AzureQueueItemLease, error) {
+	var lease AzureQueueItemLease
+	err := json.Unmarshal([]byte(leaseID), &lease)
+	if err != nil {
+		return nil, err
+	}
+	return &lease, nil
+}
+
+// Receive - Receives a collection of tasks off a given queue.
+func (s *AzqueueQueueService) Receive(options queue.ReceiveOptions) ([]queue.NitricTask, error) {
+	newErr := errors.ErrorsWithScope(
+		"AzqueueQueueService.Receive",
+		map[string]interface{}{
+			"options": options,
+		},
+	)
+
+	if err := options.Validate(); err != nil {
+		return nil, newErr(
+			codes.InvalidArgument,
+			"invalid receive options provided",
+			err,
+		)
+	}
+
+	messages := s.getMessagesUrl(options.QueueName)
+
+	ctx := context.TODO()
+	dequeueResp, err := messages.Dequeue(ctx, int32(*options.Depth), defaultVisibilityTimeout)
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"failed to received messages from the queue",
+			err,
+		)
+	}
+
+	if dequeueResp.NumMessages() == 0 {
+		return []queue.NitricTask{}, nil
+	}
+
+	// Convert the Azure Storage Queues messages into Nitric tasks
+	var tasks []queue.NitricTask
+	for i := int32(0); i < dequeueResp.NumMessages(); i++ {
+		m := dequeueResp.Message(i)
+		var nitricTask queue.NitricTask
+		err := json.Unmarshal([]byte(m.Text), &nitricTask)
+		if err != nil {
+			// TODO: append error to error list and Nack the message.
+			continue
+		}
+
+		lease := AzureQueueItemLease{
+			ID:         m.ID.String(),
+			PopReceipt: m.PopReceipt.String(),
+		}
+		leaseID, err := lease.String()
+		// This should never happen, it's a fatal error
+		if err != nil {
+			return nil, newErr(
+				codes.Internal,
+				"failed to construct queue item lease id",
+				err,
+			)
+		}
+
+		tasks = append(tasks, queue.NitricTask{
+			ID:          nitricTask.ID,
+			Payload:     nitricTask.Payload,
+			PayloadType: nitricTask.PayloadType,
+			LeaseID:     leaseID,
+		})
+	}
+
+	return tasks, nil
+}
+
+// Complete - Completes a previously popped queue item
+func (s *AzqueueQueueService) Complete(queue string, leaseId string) error {
+	newErr := errors.ErrorsWithScope(
+		"AzqueueQueueService.Complete",
+		map[string]interface{}{
+			"queue": queue,
```

Comment:
add leaseId parameter

Reply:
Could leaseId be considered sensitive? This is something only the holder should know and is not necessarily common across a stack (like queue name).

Reply:
done, might be worth considering if leaseId is sensitive information though as only its holder should know this. (not common stack knowledge like queue names for example.